### ISSUE TITLE
Introduce modes for GRM

### ISSUE THREAD

**Issue by timuthy** [gardener/gardener-resource-manager#82]
**How to categorize this issue?**

/area control-plane
/kind enhancement
/priority normal

**What would you like to be added**:
In the scope of #74, @mandelsoft @timebertt @rfranzke discussed about different use cases and modes the GRM should support in the future. We put together the following proposal:

- [ ] Add `resources.gardener.cloud/origin` to deployed resources in order to identify the responsible GRM and ManagedResource. (@mandelsoft will take over)

```
resources.gardener.cloud/origin: <Origin-ID>/<Namespace>/<Name>
```

- [ ] Add support for `resources.gardener.cloud/mode` annotation to be added to managed resources:
```
- "Ignore"
  - Removes resource from state, no action in cluster, does nothing else.
  - Use case: Move a resource from ManagedResource A -> B.

- "EnsureExists"
  - Creates resource if not available.
  - No Update.
  - Deletes resource if deleted from ManagedResource.
  - Use case: Create resource which can be adjusted by shoot owner but still can be deleted.

- "Keep"
  - Creates resource if not available
  - Updates resource
  - No Deletion
  - Use case: Create a default configuration, e.g. CoreDNS config, in shoot cluster.

- "FallBack" (default mode)
 - Creates resource if not available.
 - Updates resource if responsible as per `resources.gardener.cloud/origin` annotation.
 - Deletes resource if responsible as per `resources.gardener.cloud/origin` annotation.
 - Use case: Avoid interference of multiple GRMs targeting the same cluster, e.g. shooted seeds.
```

- [ ] Remove `"resources.gardener.cloud/ignore"` and `"resources.gardener.cloud/keep-object"` in a future version.

**Comment by timebertt** [comment]
> `- "FallBack" (default mode)`

I'm not completely sure, if `FallBack` should really be the default mode, because
- it is different to the current "default"
- it brings in a race condition / indeterministic behaviour, when we have multiple grms reconciling the same resource (e.g. the DNSEntry CRD in shooted seeds)

**Comment by timuthy** [comment]
Then we'd need a 5th mode if we don't want this to become the new default.

**Comment by timuthy** [comment]
I agree it's critical to make "Fallback" the default mode in the beginning but apart from that:

> it is different to the current "default"

I don't think it's different if you only have one GRM, isn't it?

> it brings in a race condition / indeterministic behaviour, when we have multiple grms reconciling the same resource (e.g. the DNSEntry CRD in shooted seeds)

Eventually, you have to transition to this mode since it's the only one which makes sense with multiple GRMs targeting the same cluster.

WDYT?

**Comment by rfranzke** [comment]
Another use-case (forceful shoot cluster deletion) popped up (cc @timebertt @vlerenc @dguendisch) that requires GRM to not check if the resources managed by the `ManagedResource` object are indeed deleted. If it cannot check it (maybe because the target cluster is down) then it simply continues and removes the finalizer from the `ManagedResource`.
